LFS171x/fabric-material/chaincode/tuna-app/tuna-chaincode.go

```diff
@@ -0,0 +1,222 @@
+// SPDX-License-Identifier: Apache-2.0
+
+/*
+ * Sample Chaincode based on Demonstrated Scenario
+ */
+
+package main
+
+/* Imports  
+* 4 utility libraries for handling bytes, reading and writing JSON, 
+formatting, and string manipulation  
+* 2 specific Hyperledger Fabric specific libraries for Smart Contracts  
+*/ 
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/hyperledger/fabric/core/chaincode/shim"
+	sc "github.com/hyperledger/fabric/protos/peer"
+)
+
+// Define the Smart Contract structure
+type SmartContract struct {
+}
+
+/* Define Tuna structure, with 4 properties.  
+Structure tags are used by encoding/json library
+*/
+type Tuna struct {
+	Vessel string `json:"vessel"`
+	Timestamp string `json:"timestamp"`
+	Location  string `json:"location"`
+	Holder  string `json:"holder"`
+}
+
+/*
+ * The Init method *
+ called when the Smart Contract "tuna-chaincode" is instantiated by the network
+ * Best practice is to have any Ledger initialization in separate function 
+ -- see initLedger()
+ */
+func (s *SmartContract) Init(APIstub shim.ChaincodeStubInterface) sc.Response {
+	return shim.Success(nil)
+}
+
+/*
+ * The Invoke method *
+ called when an application requests to run the Smart Contract "tuna-chaincode"
+ The app also specifies the specific smart contract function to call with args
+ */
+func (s *SmartContract) Invoke(APIstub shim.ChaincodeStubInterface) sc.Response {
+
+	// Retrieve the requested Smart Contract function and arguments
+	function, args := APIstub.GetFunctionAndParameters()
+	// Route to the appropriate handler function to interact with the ledger
+	if function == "queryTuna" {
+		return s.queryTuna(APIstub, args)
+	} else if function == "initLedger" {
+		return s.initLedger(APIstub)
+	} else if function == "recordTuna" {
+		return s.recordTuna(APIstub, args)
+	} else if function == "queryAllTuna" {
+		return s.queryAllTuna(APIstub)
+	} else if function == "changeTunaHolder" {
+		return s.changeTunaHolder(APIstub, args)
+	}
+
+	return shim.Error("Invalid Smart Contract function name.")
+}
+
+
```

Comment:
Looks like we missed a comment for this function

Reply:
done